### ISSUE TITLE
Examples of Dependency Injection in Epics?

### ISSUE THREAD

**Issue by tomwscott** [brianegan/dart_redux_epics#6]
I was wondering if there was an undocumented ability to inject dependencies into an epic. The obvious example would be for injecting a different `api` implementation for testing but I'd also like to configure different endpoints depending on whether I'm running locally or in a live like environment.

The examples in the README hint at the concept of an api - I'm just wondering if there's a way to inject it?

```
...
    .asyncMap((action) => 
      // Pseudo api that returns a Future of SearchResults
      api.search((action as PerformSearch).searchTerm)
...
```

Happy to document it if someone can point me in the right direction...

**Comment by ranquild** [comment]
Hi, we use this package with dependency injection (angular). Example:

```
@Injectable()
class TableEffects {
  final TableLoader _loader;
  final TableRouter _router;

  TableEffects(this._loader, this._router);

  Epic<TableState> getEffects() {
    return combineEpics([
      _onRouteChange,
       ...
    ]);
  }

  Stream<Object> _onRouteChange(Stream<Object> actions, EpicStore<TableState> store) {
    ...
  }
}
```

We also have store factory class to create store with injectable dependencies:

```
@Injectable()
class TableStoreFactory {
  final TableReducer _reducer;
  final TableEffects _effects;

  TableStoreFactory(this._reducer, this._effects);

  Store<TableState> getStore() {
    return new Store(
     ....
      middleware: [
        new EpicMiddleware(_effects.getEffects())
      ]
    );
  }
}

```

**Comment by brianegan** [comment]
Hey there, thanks for writing in :) 

If you're working with Flutter, there isn't a production-ready DI solution just yet, but I've heard there's one in the works. Until then, you can do it manually. You have two options: Functional style or Class-based, depending on your favorite flavor.

### Functional

```dart
Epic<AppState> createEpic(WebService service) {
  return (Stream<dynamic> actions, EpicStore<AppState> store) async* {
    service.doSomething()...
  }
}

// Usage
final myEpic = createEpic(new RealWebService());
final testEpic = createEpic(new MockWebService());
```

or with classes:

```dart
class MyEpic implements EpicClass<State> {
  final WebService service;

  MyEpic(this.service);

  @override
  Stream<dynamic> call(Stream<dynamic> actions, EpicStore<State> store) {
    service.doSomething()...
  } 
}

// Usage
final realEpic = new MyEpic(new RealWebService());
final testEpic = new TestEpic(new MockWebService());
```

**Comment by brianegan** [comment]
Fixed by #7

**Comment by leonardo2204** [comment]
@brianegan 
Sorry to revive this old thread, but how should I approach the TypedEpic type in this case?
My snippet:

```dart
selectEventAttendanceStatus(EventAttendanceRepository repository) {
  return TypedEpic((actions, store) => _searchEpic(actions, store, repository));
}
Stream<dynamic> _searchEpic(Stream<SelectEventAttendanceStatusAction> actions,
    EpicStore<AppState> state, EventAttendanceRepository repository) async* {
  await for(final action in actions) {
    try {
      yield SelectEventAttendanceStatusLoadingAction();
      final token = state.state.authState.user?.token;
      await repository.putEventStatus(action.eventId, action.status, token);
      yield SelectEventAttendanceStatusSuccessAction();
    } catch (_) {
      yield SelectEventAttendanceStatusErrorAction();
    }
  }
}
```

I'm getting: 

E/flutter (32525): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: type 'TypedEpic<dynamic, dynamic>' is not a subtype of type '(Stream<dynamic>, EpicStore<AppState>) => Stream<dynamic>'

Thanks for the great job!
